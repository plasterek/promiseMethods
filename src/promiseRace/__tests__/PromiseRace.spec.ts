import { promiseRace } from "../PromiseRace.service";

describe("promiseRace function", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  const resolvedPromise: () => Promise<unknown> = () => Promise.resolve("resolved");
  const rejectedPromise: () => Promise<unknown> = () => Promise.reject("rejected");
  const delayedResolvedPromise: () => Promise<unknown> = () =>
    new Promise((resolve) => {
      setTimeout(function () {
        resolve("resolvedWithDelay");
      }, 3000);
    });
  const delayedRejectedPromise: () => Promise<unknown> = () =>
    new Promise((reject) => {
      setTimeout(function () {
        reject("rejectedWithDelay");
      }, 3000);
    });
  describe("When first promise in given array is resolved", () => {
    it("It returns given Promise result", () => {
      //given
      jest.useFakeTimers();
      jest.spyOn(global, "setTimeout");
      const promiseArray: (() => Promise<unknown>)[] = [resolvedPromise, delayedRejectedPromise];
      //when
      const results: Promise<unknown> = promiseRace(promiseArray);
      jest.runAllTimers();
      //then
      expect(results).resolves.toMatchSnapshot();
    });
  });

  // TA NIE CHCE DZIAŁAĆ :( //
  describe("When first promise in array rejects", () => {
    it("It returns Promise with rejected value", () => {
      //given
      jest.useFakeTimers();
      jest.spyOn(global, "setTimeout");
      const promiseArray: (() => Promise<unknown>)[] = [rejectedPromise, delayedResolvedPromise];
      //when
      const results: Promise<unknown> = promiseRace(promiseArray);
      jest.runAllTimers();
      //then
      expect(results).rejects.toMatchSnapshot();
    });
  });

  describe("When given array is empty", () => {
    it("It returns Promise pending", () => {
      //given
      const promiseArray: any[] = [];
      //when
      const results: Promise<unknown> = promiseRace(promiseArray);
      //then
      expect(results).resolves.toMatchSnapshot();
    });
  });
});
