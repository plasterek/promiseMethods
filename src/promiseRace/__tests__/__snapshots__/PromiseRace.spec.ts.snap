// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`promiseRace function When first promise in array rejects It returns Promise with rejected value 1`] = `"rejected"`;

exports[`promiseRace function When first promise in given array is resolved It returns given Promise result 1`] = `"resolved"`;

exports[`promiseRace function When given array is empty It returns Promise pending 1`] = `
{
  "status": "pending",
}
`;
